GCC warning (-Wformat-truncation) for date formatting

I got the following warning with GCC 7.1.0 when compiling Duktape using -O2 optimization:
```
duk_bi_date.c: In function ‘duk_bi_date_prototype_tostring_shared’:
duk_bi_date.c:980:39: warning: ‘%02d’ directive output may be truncated writing 2 bytes into a region of size between 0 and 4 [-Wformat-truncation=]
duk_bi_date.c:980:39: note: directive argument in the range [0, 59]
In file included from duktape.h:173:0,
                 from duk_internal.h:26:
duk_config.h:2152:26: note: ‘snprintf’ output between 7 and 11 bytes into a destination of size 8
 #define DUK_SNPRINTF     snprintf
duk_bi_date.c:980:4: note: in expansion of macro ‘DUK_SNPRINTF’
```

--- a/RELEASES.rst
+++ b/RELEASES.rst
@@ -2994,6 +2994,11 @@ Planned
 * Fix duk_debug.js --source-dirs scanning for file dropdown; the dropdown was
   empty (GH-1580)
 
+* Fix some compile warnings for duk_error() and duk_throw() in duk_cmdline.c,
+  module-duktape, module-node, and duk-v1-compat extras (GH-1646)
+
+* Fix harmless GCC 7 -Wextra warning in Date built-in (GH-1646)
+
 * Improve support for old MSVC versions without __pragma(), long long, and
   LL/ULL constants (GH-1559, GH-1562)
 

--- a/examples/cmdline/duk_cmdline.c
+++ b/examples/cmdline/duk_cmdline.c
@@ -277,12 +277,12 @@ static duk_ret_t wrapped_compile_execute(duk_context *ctx, void *udata) {
 
 		f = fopen(fnbuf, "wb");
 		if (!f) {
-			duk_error(ctx, DUK_ERR_ERROR, "failed to open bytecode output file");
+			(void) duk_generic_error(ctx, "failed to open bytecode output file");
 		}
 		wrote = fwrite(bc_ptr, 1, (size_t) bc_len, f);  /* XXX: handle partial writes */
 		(void) fclose(f);
 		if (wrote != bc_len) {
-			duk_error(ctx, DUK_ERR_ERROR, "failed to write all bytecode");
+			(void) duk_generic_error(ctx, "failed to write all bytecode");
 		}
 
 		return 0;  /* duk_safe_call() cleans up */
@@ -889,22 +889,22 @@ static duk_ret_t fileio_read_file(duk_context *ctx) {
 	fn = duk_require_string(ctx, 0);
 	f = fopen(fn, "rb");
 	if (!f) {
-		duk_error(ctx, DUK_ERR_TYPE_ERROR, "cannot open file %s for reading, errno %ld: %s",
-		          fn, (long) errno, strerror(errno));
+		(void) duk_type_error(ctx, "cannot open file %s for reading, errno %ld: %s",
+		                      fn, (long) errno, strerror(errno));
 	}
 
 	rc = fseek(f, 0, SEEK_END);
 	if (rc < 0) {
 		(void) fclose(f);
-		duk_error(ctx, DUK_ERR_TYPE_ERROR, "fseek() failed for %s, errno %ld: %s",
-		          fn, (long) errno, strerror(errno));
+		(void) duk_type_error(ctx, "fseek() failed for %s, errno %ld: %s",
+		                      fn, (long) errno, strerror(errno));
 	}
 	len = (size_t) ftell(f);
 	rc = fseek(f, 0, SEEK_SET);
 	if (rc < 0) {
 		(void) fclose(f);
-		duk_error(ctx, DUK_ERR_TYPE_ERROR, "fseek() failed for %s, errno %ld: %s",
-		          fn, (long) errno, strerror(errno));
+		(void) duk_type_error(ctx, "fseek() failed for %s, errno %ld: %s",
+		                      fn, (long) errno, strerror(errno));
 	}
 
 	buf = (char *) duk_push_fixed_buffer(ctx, (duk_size_t) len);
@@ -913,14 +913,14 @@ static duk_ret_t fileio_read_file(duk_context *ctx) {
 		got = fread((void *) (buf + off), 1, len - off, f);
 		if (ferror(f)) {
 			(void) fclose(f);
-			duk_error(ctx, DUK_ERR_TYPE_ERROR, "error while reading %s", fn);
+			(void) duk_type_error(ctx, "error while reading %s", fn);
 		}
 		if (got == 0) {
 			if (feof(f)) {
 				break;
 			} else {
 				(void) fclose(f);
-				duk_error(ctx, DUK_ERR_TYPE_ERROR, "error while reading %s", fn);
+				(void) duk_type_error(ctx, "error while reading %s", fn);
 			}
 		}
 		off += got;
@@ -943,7 +943,7 @@ static duk_ret_t fileio_write_file(duk_context *ctx) {
 	fn = duk_require_string(ctx, 0);
 	f = fopen(fn, "wb");
 	if (!f) {
-		duk_error(ctx, DUK_ERR_TYPE_ERROR, "cannot open file %s for writing, errno %ld: %s",
+		(void) duk_type_error(ctx, "cannot open file %s for writing, errno %ld: %s",
 		          fn, (long) errno, strerror(errno));
 	}
 
@@ -954,11 +954,11 @@ static duk_ret_t fileio_write_file(duk_context *ctx) {
 		got = fwrite((const void *) (buf + off), 1, len - off, f);
 		if (ferror(f)) {
 			(void) fclose(f);
-			duk_error(ctx, DUK_ERR_TYPE_ERROR, "error while writing %s", fn);
+			(void) duk_type_error(ctx, "error while writing %s", fn);
 		}
 		if (got == 0) {
 			(void) fclose(f);
-			duk_error(ctx, DUK_ERR_TYPE_ERROR, "error while writing %s", fn);
+			(void) duk_type_error(ctx, "error while writing %s", fn);
 		}
 		off += got;
 	}

--- a/extras/duk-v1-compat/duk_v1_compat.c
+++ b/extras/duk-v1-compat/duk_v1_compat.c
@@ -60,7 +60,7 @@ const char *duk_push_string_file_raw(duk_context *ctx, const char *path, duk_uin
 	if (flags & DUK_STRING_PUSH_SAFE) {
 		duk_push_undefined(ctx);
 	} else {
-		duk_error(ctx, DUK_ERR_TYPE_ERROR, "read file error");
+		(void) duk_type_error(ctx, "read file error");
 	}
 	return NULL;
 }

--- a/extras/module-duktape/duk_module_duktape.c
+++ b/extras/module-duktape/duk_module_duktape.c
@@ -180,7 +180,7 @@ static void duk__resolve_module_id(duk_context *ctx, const char *req_id, const c
 	return;
 
  resolve_error:
-	duk_error(ctx, DUK_ERR_TYPE_ERROR, "cannot resolve module id: %s", (const char *) req_id);
+	(void) duk_type_error(ctx, "cannot resolve module id: %s", (const char *) req_id);
 }
 
 /* Stack indices for better readability. */
@@ -432,7 +432,7 @@ static duk_ret_t duk__require(duk_context *ctx) {
  delete_rethrow:
 	duk_dup(ctx, DUK__IDX_RESOLVED_ID);
 	duk_del_prop(ctx, DUK__IDX_MODLOADED);  /* delete Duktape.modLoaded[resolved_id] */
-	duk_throw(ctx);  /* rethrow original error */
+	(void) duk_throw(ctx);  /* rethrow original error */
 	return 0;  /* not reachable */
 }
 

--- a/extras/module-duktape/test.c
+++ b/extras/module-duktape/test.c
@@ -12,7 +12,7 @@ static duk_ret_t handle_assert(duk_context *ctx) {
 	if (duk_to_boolean(ctx, 0)) {
 		return 0;
 	}
-	duk_error(ctx, DUK_ERR_ERROR, "assertion failed: %s", duk_safe_to_string(ctx, 1));
+	(void) duk_generic_error(ctx, "assertion failed: %s", duk_safe_to_string(ctx, 1));
 	return 0;
 }
 

--- a/extras/module-node/duk_module_node.c
+++ b/extras/module-node/duk_module_node.c
@@ -133,7 +133,7 @@ static duk_ret_t duk__handle_require(duk_context *ctx) {
 		duk_pop(ctx);
 	} else {
 		duk__del_cached_module(ctx, id);
-		(void) duk_error(ctx, DUK_ERR_TYPE_ERROR, "invalid module load callback return value");
+		(void) duk_type_error(ctx, "invalid module load callback return value");
 	}
 
 	/* fall through */

--- a/extras/module-node/test.c
+++ b/extras/module-node/test.c
@@ -41,7 +41,7 @@ static duk_ret_t cb_load_module(duk_context *ctx) {
 	} else if (strcmp(module_id, "shebang.js") == 0) {
 		duk_push_string(ctx, "#!ignored\nexports.foo = 123; exports.bar = 234;");
 	} else {
-		duk_error(ctx, DUK_ERR_TYPE_ERROR, "cannot find module: %s", module_id);
+		(void) duk_type_error(ctx, "cannot find module: %s", module_id);
 	}
 
 	return 1;
@@ -56,7 +56,7 @@ static duk_ret_t handle_assert(duk_context *ctx) {
 	if (duk_to_boolean(ctx, 0)) {
 		return 0;
 	}
-	duk_error(ctx, DUK_ERR_ERROR, "assertion failed: %s", duk_safe_to_string(ctx, 1));
+	(void) duk_generic_error(ctx, "assertion failed: %s", duk_safe_to_string(ctx, 1));
 	return 0;
 }
 

--- a/src-input/duk_bi_date.c
+++ b/src-input/duk_bi_date.c
@@ -974,13 +974,23 @@ DUK_LOCAL void duk__format_parts_iso8601(duk_int_t *parts, duk_int_t tzoffset, d
 		/* tzoffset seconds are dropped; 16 bits suffice for
 		 * time offset in minutes
 		 */
+		const char *fmt;
+		duk_small_int_t tmp, arg_hours, arg_minutes;
+
 		if (tzoffset >= 0) {
-			duk_small_int_t tmp = tzoffset / 60;
-			DUK_SNPRINTF(tzstr, sizeof(tzstr), "+%02d:%02d", (int) (tmp / 60), (int) (tmp % 60));
+			tmp = tzoffset;
+			fmt = "+%02d:%02d";
 		} else {
-			duk_small_int_t tmp = -tzoffset / 60;
-			DUK_SNPRINTF(tzstr, sizeof(tzstr), "-%02d:%02d", (int) (tmp / 60), (int) (tmp % 60));
+			tmp = -tzoffset;
+			fmt = "-%02d:%02d";
 		}
+		tmp = tmp / 60;
+		arg_hours = tmp / 60;
+		arg_minutes = tmp % 60;
+		DUK_ASSERT(arg_hours <= 24);  /* Even less is actually guaranteed for a valid tzoffset. */
+		arg_hours = arg_hours & 0x3fU;  /* For [0,24] this is a no-op, but fixes GCC 7 warning, see https://github.com/svaarala/duktape/issues/1602. */
+
+		DUK_SNPRINTF(tzstr, sizeof(tzstr), fmt, (int) arg_hours, (int) arg_minutes);
 		tzstr[sizeof(tzstr) - 1] = (char) 0;
 	} else {
 		tzstr[0] = DUK_ASC_UC_Z;

